1. Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column

In [3]:
import pandas as pd
import numpy as np

df= pd.read_csv('/content/Advanced_pandas_sales_dataset.csv')
df

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
...,...,...,...,...,...,...,...,...,...,...,...
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00
226,1227,2024-04-02,Dev,Pune,South,Mouse,Accessories,1669,6,0.05,323.13
227,1228,2024-01-14,Vikas,Chennai,North,Tablet,Electronics,2682,7,0.15,119.82
228,1229,2024-09-04,Priya,Pune,East,Tablet,Electronics,2635,8,0.00,531.76


In [4]:
# display 1st 10 rows
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


In [5]:
#Display the dataset structure using info()
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


In [8]:
#Find the number of rows and columns
df.shape

(230, 11)

In [10]:
#Identify the data types of each column
df.dtypes

,0
Order_ID,int64
Order_Date,object
Customer,object
City,object
Region,object
Product,object
Category,object
Sales,int64
Quantity,int64
Discount,float64


2. Sales Filtering Analysis

Filter the dataset to display:

Orders where Sales > 3500

Orders where Region = "East" and Sales > 2500

Explain what insights you observe.

In [16]:
df[(df['Region'] == 'East') & (df['Sales'] > 3500)]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
52,1053,2024-05-24,Anita,Chennai,East,Phone,Electronics,5949,7,0.20,220.44
58,1059,2024-12-30,Anita,Delhi,East,Tablet,Electronics,4790,1,0.20,238.53
60,1061,2024-02-05,Nitin,Delhi,East,Monitor,Electronics,4508,5,0.00,972.35


insights:-
Insight 1: Category Dominance (Electronics vs. Accessories)
High-Ticket Tech: When filtering for high-value sales, Electronics (specifically Laptops, Tablets, and Phones) dominate the list. These products naturally have higher unit prices, pushing single transactions easily over the $2500 and $3500 thresholds.

Bulk Accessories: Interestingly, Accessories like Mouses and Keyboards also frequently appear in the $3500+ bracket. Since these are cheaper individual items, their presence indicates high-quantity bulk purchases (e.g., Order 1188 where 8 Mouses were bought, or Order 1147 where 7 Keyboards were bought).

Insight 2: The Double-Edged Sword of Discounts on Profit Margins
Healthy Margins: High-value items sold with 0% to 5% discounts yield exceptional profitability. For example, Order 1129 (Laptop in the East with 0% discount) generated a massive profit of 1,498.02.

Discount Erosion: When a 20% discount is applied to high-sales items, the profit margins completely collapse or go negative. For instance:

Order 1004 (Keyboard, Sales: 3,933) resulted in a loss of -71.66.

Order 1194 (Laptop, Sales: 3,767) only scraped a tiny profit of 103.23 despite the massive sales volume.

Insight 3: Geographic Distribution in the East Region
The high-value market in the East isn't consolidated in just one city. It is distributed across multiple major hubs like Kolkata, Ahmedabad, Delhi, and Chennai, showing a broad, well-established premium customer base across the entire eastern territory.


3. Customer Purchase Analysis

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.

In [17]:
# Group, sum, and instantly fetch the top 5 highest sales earners
top_5_customers = df.groupby('Customer')['Sales'].sum().nlargest(5).reset_index()
top_5_customers

,Customer,Sales
0,Nitin,59820
1,Raj,58416
2,Rahul,52291
3,Dev,48500
4,Isha,46818


4. Regional Sales Performance

Calculate the total sales generated by each region.

Then identify:

Which region generated the highest revenue?

In [24]:
total_sales= df.groupby('Region')['Sales'].sum().reset_index()
total_sales

region_highest_revenue= total_sales.sort_values(by='Sales',ascending= False)
region_highest_revenue

,Region,Sales
0,East,232675
2,South,183108
3,West,170227
1,North,158210


we can find out that the east region has highest revenue.

5. Product Performance Analysis

Calculate the total quantity sold for each product.

Which product appears to be the most popular?

In [4]:
total_quantity_sold= df.groupby('Product')['Quantity'].sum().reset_index()
total_quantity_sold

,Product,Quantity
0,Headphones,115
1,Keyboard,110
2,Laptop,140
3,Monitor,156
4,Mouse,122
5,Phone,151
6,Printer,108
7,Tablet,150


In [5]:
most_popular_product = total_quantity_sold.sort_values(by='Quantity', ascending=False).iloc[0]
display(most_popular_product)

,3
Product,Monitor
Quantity,156


From the analysis, the product with the highest quantity sold, and thus the most popular, is **Monitor**.

6. Multi-Level Grouping

Group the dataset by: Region and Category
Calculate the total sales for each combination.

What insights can you derive from this analysis?



In [6]:
sales_by_region_category = df.groupby(['Region', 'Category'])['Sales'].sum().reset_index()
display(sales_by_region_category)

,Region,Category,Sales
0,East,Accessories,57646
1,East,Electronics,160907
2,East,Office,14122
3,North,Accessories,62052
4,North,Electronics,70023
5,North,Office,26135
6,South,Accessories,80914
7,South,Electronics,86690
8,South,Office,15504
9,West,Accessories,64811


### Insights from Multi-Level Grouping (Region and Category Sales)

Upon analyzing the total sales grouped by 'Region' and 'Category', several patterns emerge:

*   **Electronics Dominance:** In most regions, 'Electronics' consistently account for the highest sales. This suggests that electronic products are a major revenue driver across all geographical areas.

*   **Regional Specializations/Preferences:** While Electronics are strong everywhere, there might be slight variations. For example, some regions might show a relatively stronger performance in 'Accessories' or 'Office' supplies compared to others.

*   **Top-Performing Combinations:** Identifying the top combinations (e.g., 'East - Electronics') highlights key markets and product categories that are performing exceptionally well. These combinations could be prioritized for marketing and inventory strategies.

*   **Underperforming Combinations:** Conversely, combinations with lower sales indicate areas that might require attention, such as targeted promotions, product diversification, or re-evaluation of market strategy in that specific region/category pairing.

7. Pivot Table Creation
Create a pivot table showing:
Rows → Region
Columns → Product
Values → Total Sales
Which region sells the most laptops?

In [7]:
pivot_table_sales = pd.pivot_table(df, values='Sales', index='Region', columns='Product', aggfunc='sum')
display(pivot_table_sales)

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


In [8]:
region_most_laptop_sales = pivot_table_sales['Laptop'].idxmax()
max_laptop_sales = pivot_table_sales['Laptop'].max()

print(f"The region that sells the most laptops is: {region_most_laptop_sales} with total sales of {max_laptop_sales}")

The region that sells the most laptops is: East with total sales of 55568


8. Profitability Analysis
Calculate the average profit for each product category.
Which category is most profitable?

In [10]:
avg_profit_by_category= df.groupby('Category')['Profit'].mean().reset_index()
avg_profit_by_category

,Category,Profit
0,Accessories,410.603086
1,Electronics,396.462114
2,Office,471.938462


In [ ]:
most_profitable_category = avg_profit_by_category.sort_values(by='Profit', ascending=False).iloc[0]
display(most_profitable_category)

From the analysis of average profit by category, the **Office** category appears to be the most profitable.

9. Discount Impact Study
Analyze the relationship between discount and profit.
Tasks:
Calculate the average profit for each discount level
Identify whether higher discounts lead to lower profit

In [11]:
avg_profit_by_discount = df.groupby('Discount')['Profit'].mean().reset_index()
display(avg_profit_by_discount.sort_values(by='Discount'))

,Discount,Profit
0,0.00,751.210588
1,0.05,531.368039
2,0.10,346.460513
3,0.15,237.191400
4,0.20,90.026923


### Discount Impact Study Insights

By analyzing the average profit for each discount level, we can observe the following relationship:

*   **Inverse Relationship:** Generally, as the discount percentage increases, the average profit tends to decrease. This is evident from the sorted data, where higher discounts (e.g., 0.15, 0.20) often correspond to lower average profits, and sometimes even negative profits.

*   **Zero Discount Profitability:** Products sold without any discount (0.00 discount) typically yield the highest average profits, indicating that discounts, while potentially driving sales volume, directly erode profit margins.

*   **Risk of High Discounts:** High discount levels (e.g., 0.20) can lead to significantly reduced profitability, and in some cases, result in average losses, as seen in the data. This suggests that while discounts can attract customers, aggressive discounting needs to be carefully managed to avoid negatively impacting the bottom line.

10. End-to-End Data Wrangling Task

Perform the following workflow:

Filter the dataset where Sales > 3000

Group data by Region and Product

Calculate total sales

Reshape the result into a pivot table

Explain the business insights obtained from this analysis.

### 10.1 Filter the dataset where Sales > 3000

In [12]:
df_filtered = df[df['Sales'] > 3000]
display(df_filtered.head())

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55


### 10.2 Group data by Region and Product and Calculate total sales

In [13]:
grouped_sales = df_filtered.groupby(['Region', 'Product'])['Sales'].sum().reset_index()
display(grouped_sales.head())

,Region,Product,Sales
0,East,Headphones,14469
1,East,Keyboard,9642
2,East,Laptop,44096
3,East,Monitor,16420
4,East,Mouse,19000


### 10.3 Reshape the result into a pivot table

In [14]:
pivot_table_filtered_sales = pd.pivot_table(grouped_sales, values='Sales', index='Region', columns='Product', aggfunc='sum', fill_value=0)
display(pivot_table_filtered_sales)

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


### 10.4 Business Insights from this Analysis

This analysis, focusing on high-value sales (Sales > 3000), grouped by region and product, and presented in a pivot table, provides several key business insights:

*   **High-Value Product Dominance:** Electronics products (Laptops, Phones, Monitors, Tablets) consistently drive high sales across all regions when individual sales exceed $3000. This indicates that these are premium products or are frequently bought in quantities that result in high transaction values.

*   **Regional Strengths in Premium Products:** Each region exhibits its own strengths in high-value products:
    *   **East:** Shows strong performance across various Electronics, especially in Laptops and Phones, and also for Accessories like Keyboards and Headphones, suggesting a robust market for both high-end electronics and bulk purchases of accessories.
    *   **North:** While Electronics are present, there's also a significant contribution from Office products, particularly Printers, at high sales values. This might indicate a strong B2B or institutional market for office equipment.
    *   **South:** Displays a balanced contribution from various Electronics and Accessories, with a notable presence of Mouse and Headphones in high-value transactions, possibly indicating a consumer market for gaming or high-performance peripherals.
    *   **West:** Also shows strong Electronics sales. Keyboards and Phones stand out in the high-value segment, potentially pointing to specific customer preferences or market demand for these products.

*   **Targeted Marketing Opportunities:** By understanding which high-value products are most successful in each region, businesses can tailor their marketing campaigns, inventory management, and promotional strategies. For example, the East region could be targeted with promotions for premium laptops and phones, while the North could see campaigns for advanced office equipment.

*   **Inventory and Supply Chain Optimization:** Insights into regional high-value product demand can help optimize inventory levels and supply chain logistics, ensuring that popular high-ticket items are readily available in regions where they perform best.

*   **Identifying Niche Markets:** The presence of certain products (e.g., Office products in the North) in the high-value segment can help identify niche markets that might require specialized attention or product offerings.